packages/electric-db-collection/src/electric.ts

```diff
@@ -239,67 +406,157 @@ export function electricCollectionOptions(
+      // This will be handled by the sync configuration
+      const checkMatch = (message: Message<any>) => {
+        if (matchFn(message)) {
+          debug(`awaitMatch found matching message, waiting for up-to-date`)
+          // Mark as matched but don't resolve yet - wait for up-to-date
+          pendingMatches.setState((current) => {
+            const newMatches = new Map(current)
+            const existing = newMatches.get(matchId)
+            if (existing) {
+              newMatches.set(matchId, { ...existing, matched: true })
+            }
+            return newMatches
+          })
+          return true
         }
+        return false
+      }
 
-        // Handle both single txid and array of txids
-        if (Array.isArray(txid)) {
-          await Promise.all(txid.map((id) => awaitTxId(id)))
-        } else {
-          await awaitTxId(txid)
+      // Check against current batch messages first to handle race conditions
+      for (const message of currentBatchMessages.state) {
+        if (checkMatch(message)) {
+          debug(
+            `awaitMatch found immediate match in current batch, waiting for up-to-date`
+          )
+          // Mark as matched and register for up-to-date resolution
+          pendingMatches.setState((current) => {
+            const newMatches = new Map(current)
+            newMatches.set(matchId, {
+              matchFn: checkMatch,
+              resolve,
+              reject,
+              timeoutId,
+              matched: true, // Already matched
+            })
+            return newMatches
+          })
+          return
         }
+      }
+
+      // Store the match function for the sync process to use
+      // We'll add this to a pending matches store
+      pendingMatches.setState((current) => {
+        const newMatches = new Map(current)
+        newMatches.set(matchId, {
+          matchFn: checkMatch,
+          resolve,
+          reject,
+          timeoutId,
+          matched: false,
+        })
+        return newMatches
+      })
+    })
+  }
+
+  /**
+   * Wait for a fixed timeout (void strategy)
+   * @param timeout Timeout in milliseconds (defaults to 3000ms for void strategy)
+   * @returns Promise that resolves after the timeout
+   */
+  const awaitVoid = async (timeout: number = 3000): Promise<boolean> => {
+    debug(`awaitVoid called with timeout %dms`, timeout)
+    return new Promise((resolve) => {
+      setTimeout(() => {
+        debug(`awaitVoid completed after %dms`, timeout)
+        resolve(true)
+      }, timeout)
+    })
+  }
+
+  /**
+   * Process matching strategy and wait for synchronization
+   */
+  const processMatchingStrategy = async (
+    result: MatchingStrategy<any>
+  ): Promise<void> => {
+    // Check for txid strategy (backward compatible)
+    if (`txid` in result) {
+      // Handle both single txid and array of txids
+      if (Array.isArray(result.txid)) {
+        await Promise.all(result.txid.map((id) => awaitTxId(id)))
```

Comment:
nit: we can shorten `result.txid.map((id) => awaitTxId(id))` to `result.txid.map(awaitTxId)`